contracts/nftauction.js

```diff
@@ -0,0 +1,597 @@
+    auctionId,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (api.assert(isSignedWithActiveKey === true, 'you must use a custom_json signed with your active key')
+    && api.assert(auctionId && typeof auctionId === 'string', 'invalid params')) {
+    const auction = await api.db.findOne('auctions', { auctionId });
+
+    if (api.assert(auction, 'auction does not exist or has been expired')) {
+      const {
+        priceSymbol,
+        currentLead,
+        expiryTimestamp,
+        lastLeadUpdate,
+      } = auction;
+
+      // find if the account has any bid in this auction
+      const bidIndex = auction.bids.findIndex(el => el.account === api.sender);
+      const bid = auction.bids[bidIndex];
+
+      const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+      const timestamp = blockDate.getTime();
+
+      if (api.assert(bid, 'you don not have a bid in this auction')
+        && api.assert(expiryTimestamp >= timestamp, 'auction has been expired')) {
+        const params = await api.db.findOne('params', {});
+        const timeRemaining = api.BigNumber(lastLeadUpdate)
+          .plus(params.expiryTimeMillis)
+          .minus(timestamp);
+        const timeRemainingExpire = api.BigNumber(expiryTimestamp).minus(timestamp);
+
+        // do not cancel bid if auction is about to settle
+        if (api.assert(timeRemaining.gt(params.cancelLockTimeMillis)
+          && timeRemainingExpire.gt(params.cancelLockTimeMillis), 'can not cancel bid when auction is about to settle')) {
+          // remove this bid from the auction
+          auction.bids.splice(bidIndex, 1);
+
+          if (currentLead === bidIndex) {
+            let largestBid = '0';
+            let largestBidIndex = null;
+            for (let i = 0; i < auction.bids.length; i += 1) {
+              const { bid: quantity } = auction.bids[i];
+              if (api.BigNumber(quantity).gt(largestBid)) {
+                largestBid = quantity;
+                largestBidIndex = i;
+              }
+            }
+
+            auction.currentLead = largestBidIndex;
+          } else if (currentLead > bidIndex) {
+            // re-assign lead if a lower indexed bid is removed
+            auction.currentLead -= 1;
+          }
+
+          await api.transferTokens(api.sender, priceSymbol, bid.bid, 'user');
+
+          await api.db.update('auctions', auction);
+
+          api.emit('cancelBid', {
+            auctionId,
+            ...bid,
+          });
+        }
+      }
+    }
+  }
+};
+
+actions.updateAuctions = async () => {
+  if (!api.assert(api.sender === 'null', 'not authorized')) return;
+
+  const params = await api.db.findOne('params', {});
+
+  const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
+  const timestamp = blockDate.getTime();
+
+  const lastValidLead = timestamp - params.expiryTimeMillis;
+
+  const auctionsToSettle = await api.db.find('auctions',
+    {
+      $or: [
+        { lastLeadUpdate: { $lte: lastValidLead } },
```

Comment:
for these, you should add indexes for 'lastLeadUpdate'  and 'expiryTimestamp' even though you do not sort on them